### PR TITLE
Code for dev, staging and prod AWS SSM and AWS config lambda

### PR DESCRIPTION
**Pull request by wshands**
This PR introduces code to send messages to Slack for dev, staging and prod. The lambda is different from what is currently on dev.dockstore.net in that this staging code uses AWS managed keys to encrypt the environment variables at rest instead of using a customer managed key. This simplifies the code. It's unclear to me why we need to use a customer managed key.

Strangely, the lambda currently on dev.dockstore.net handles mail attachments; but that is not reflected in the lambda code checked in here for develop. It's not clear to me if this code is needed.

Additionally, there are output-template.yml, template.yml files that do not match this new index.js code with respect to the AWS managed key use. It's unclear to me how these are generated.

This code is enabled with the CloudFormation template from PR https://github.com/dockstore/dockstore-deploy/pull/174.

### CODE REVIEW

**Comment by coverbeck** [comment]
> Ultimately this code will be replaced with a CloudFormation template.

Code from this repo would still be used. CloudFormation would only deploy it.

**Comment by coverbeck** [comment]
> This PR introduces code to send messages to Slack for staging.dockstore.org. The lambda is different from what is currently on dev.dockstore.net in that this staging code uses AWS managed keys to encrypt the environment variables at rest instead of using a customer managed key. This simplifies the code. It's unclear to me why we need to use a customer managed key.

Going to defer to Denis on that one.
 
> Strangely, the lambda currently on dev.dockstore.net handles mail attachments; but that is not reflected in the lambda code checked in here for develop. It's not clear to me if this code is needed.

Pretty sure the code in prod would send out an email, although I think we just disabled the alarm that triggers it. I'm just going off your comments here, but I don't think we should remove it.

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (coverbeck) [inline cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:127]: You could probably just leave this uncommented.

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (coverbeck) [inline cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:73]: Was it like this before (foo)?

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (coverbeck) [inline cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:82]: This line confuses me... The code to check for SSM events was already committed? And that is what was deployed to dev? But then you had to make additional changes to deploy it to staging?

**Review by coverbeck — COMMENTED** [review]
I'm a little confused. Seems like all the changes are because of the key? And email? I guess it would be good to know if the AWS key approach is OK or not before proceeding. Will wait for Denis.

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:82]: The code editor wanted a semicolon at the end of the line; I guess I missed that before. The missing semicolon doesn't cause a failure.

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:73]: Yes it is that way on dev.dockstore.net

**Comment by denis-yuen** [comment]
>  It's unclear to me why we need to use a customer managed key.
I think this customer managed keys were requested by UCSC ITS the first time we engaged that group. So I left it in this when I used a template to create this lambda. I think I'd prefer not customer managed if we don't need it.

**Comment by denis-yuen** [comment]
> Strangely, the lambda currently on dev.dockstore.net handles mail attachments; but that is not reflected in the lambda code checked in here for develop. It's not clear to me if this code is needed.

I think this was for our support email, should be kept/maintained.

**Review by denis-yuen — COMMENTED** [review]
Can remove customer key if not needed for FedRAMP effort. 
Not sure about email attachment behavior, might be failed attempt at sending email to slack

**Comment by wshands** [comment]
> > Strangely, the lambda currently on dev.dockstore.net handles mail attachments; but that is not reflected in the lambda code checked in here for develop. It's not clear to me if this code is needed.
> 
> I think this was for our support email, should be kept/maintained.
> 
> Edit: maybe not, looks like its currently hooked up to send emails to us and not to #slack directly https://oicr.slack.com/archives/C011TT6UKQR/p1607702040162900

I will remove email handling from this Lambda code.

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js]: now uncommented

**Comment by wshands** [comment]
I removed email handling

**Review by coverbeck — COMMENTED** [review]
Lambda code seems fine off-hand. We can always restore the email.

The PR title and description are confusing, because this is not for staging.dockstore.org, but for all environments.

**Review by denis-yuen — APPROVED** [review]

**Comment by wshands** [comment]
I added code to get the EC2 instance name. This was more complicated than I anticipated, and in IMHO this lambda code code be written more clearly in Python.

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (garyluu) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:37]: Interesting, I ran prettier and eslint (with the existing config) on this repo. It seems not have https://eslint.org/docs/rules/prefer-const set

**Review by garyluu — COMMENTED** [review]
In the event we want to run the existing eslint.js on this file, additional https://eslint.org/docs/rules/no-prototype-builtins should probably be avoided.

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js]: fixed

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js]: done

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js]: fixed

**Review by garyluu — APPROVED** [review]

**Comment by david4096** [comment]
> In the event we want to run the existing eslint.js on this file, additional https://eslint.org/docs/rules/no-prototype-builtins should probably be avoided.

I think we do? Do we need to add hooks for linting to happen?

**Comment by garyluu** [comment]
> > In the event we want to run the existing eslint.js on this file, additional https://eslint.org/docs/rules/no-prototype-builtins should probably be avoided.
> 
> I think we do? Do we need to add hooks for linting to happen?

https://github.com/dockstore/lambda/blob/develop/.github/workflows/maven.yml#L18-L19

**Review by GFJHogue — APPROVED** [review]

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (coverbeck) [inline cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:40]: see comment on line 49

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (coverbeck) [inline cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:49]: If you get an error as on line 40, or you don't find tagInstanceNameKey; you're not going to send a message to Slack. I think you still should.

Also, if you're going to keep this `if`, then you can move line 48 inside it and not test for tagInstanceNameKey being set.

Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js:49]: fixed

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js]: fixed

File: cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js
Comment (wshands) [reply cloud-watch-to-slack-testing-58006136-c30b-43d1-880c-7b36f860e7c8/deployment/index.js]: fixed

**Review by denis-yuen — CHANGES_REQUESTED** [review]
Reminder, you'll want to hook up linting as in https://github.com/dockstore/lambda/blob/develop/.github/workflows/maven.yml#L18 before merging

**Comment by wshands** [comment]
added linting

**Review by coverbeck — APPROVED** [review]

**Comment by wshands** [comment]
Hi @denis-yuen would you approve please?

**Review by denis-yuen — APPROVED** [review]

**Pull request merged**
